src/btree/bt_read.c

```diff
@@ -184,6 +184,281 @@ __bt_reconstruct_delta(WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *delta)
+    /* Overflow keys are not supported. */
+    default:
+        WT_RET(__wt_illegal_value(session, delta->key.type));
+    }
+
+    switch (value_type) {
+    case WT_CELL_ADDR_INT:
+        F_SET(ref, WT_REF_FLAG_INTERNAL);
+        break;
+    case WT_CELL_ADDR_LEAF:
+    case WT_CELL_ADDR_LEAF_NO:
+        F_SET(ref, WT_REF_FLAG_LEAF);
+        break;
+    case WT_CELL_ADDR_DEL:
+    /* Fast truncated pages are not supported. */
+    default:
+        WT_RET(__wt_illegal_value(session, delta->value.type));
+    }
+
+    if (base)
+        ref->addr = base_val->cell;
+    else
+        WT_RET(__wt_memdup(session, delta->value.data, delta->value.size, &ref->addr));
+    return (0);
+}
+
+/*
+ * __bt_merge_internal_delta_with_base_image --
+ *     Merge internal delta with the base image
+ */
+static int
+__bt_merge_internal_delta_with_base_image(WT_SESSION_IMPL *session, WT_REF *ref,
+  WT_CELL_UNPACK_DELTA_INT **delta, size_t delta_entries, WT_REF ***refsp, size_t *ref_entriesp,
+  size_t *incr)
+{
+    WT_CELL_UNPACK_ADDR *base, *base_key, *base_val;
+    WT_DECL_RET;
+    WT_ITEM base_key_buf, delta_key_buf;
+    WT_PAGE *page;
+    WT_REF **refs;
+    size_t i, j, k, base_entries, estimated_entries, final_entries;
+    int cmp;
+
+    final_entries = i = j = k = 0;
+    page = ref->page;
+    base_entries = (size_t)page->dsk->u.entries;
+
+    WT_ASSERT(session, base_entries != 0 && delta_entries != 0);
+
+    /* Unpack the base image. */
+    WT_ERR(__wt_calloc_def(session, base_entries, &base));
+    WT_CELL_FOREACH_ADDR (session, page->dsk, base[k]) {
+        k++;
+    }
+    WT_CELL_FOREACH_END;
+
+    /* Create a new ref array which would be the merged refs from the base image and the deltas. */
+    estimated_entries = (base_entries / 2) + delta_entries + 1;
+    WT_ERR(__wt_calloc_def(session, estimated_entries, refsp));
+    refs = *refsp;
+
+    /* Perform a merge sort between the base image and the deltas. */
+    while (i < base_entries && j < delta_entries) {
+        /* Compare the keys of the base image and the delta. */
+        base_key_buf.data = base[i].data;
+        base_key_buf.size = base[i].size;
+        delta_key_buf.data = delta[j]->key.data;
+        delta_key_buf.size = delta[j]->key.size;
+        WT_ERR(__wt_compare(session, S2BT(session)->collator, &base_key_buf, &delta_key_buf, &cmp));
+
+        if (cmp < 0) {
+            base_key = &base[i++];
+            base_val = &base[i++];
+            WT_ERR(__bt_build_ref(
+              session, ref, base_key, base_val, NULL, true, &refs[final_entries++], incr));
+        } else if (cmp >= 0) {
+            if (!F_ISSET(delta[j], WT_DELTA_INT_IS_DELETE))
+                __bt_build_ref(
+                  session, ref, NULL, NULL, delta[j], false, &refs[final_entries++], incr);
+            if (cmp == 0)
+                i += 2; /* Skip the current key and value. */
+            j++;
+        }
+    }
+
+    for (; i < base_entries;) {
```

Comment:
May be we can add assert that j == delta_entries?

Reply:
updates in [6021346](https://github.com/wiredtiger/wiredtiger/pull/11627/commits/60213461c2b487f3e6f92235c57d830d8007741d)